src/third_party/libkmip-0ecda33/kmippp/kmippp.cpp

```diff
@@ -15,11 +17,142 @@
+/// In our case, the password is known in advance so that the callback only
+/// copies it from `userdata` to `buffer`.
+///
+/// @param buffer buffer the password should be read to
+/// @param buffer_size the size of the buffer
+/// @param rwflag set to 0 by `ssl` for decryption and to 1 for encryption;
+///               unused in this particular callback
+/// @param userdata user-provided callback data; in this callback it must be
+///                 a pointer to the password in the form of `const std::string`.
+///
+/// @returns The password length or -1 in case of error.
+int passwd_cb(char* buffer, int buffer_size, [[maybe_unused]] int rwflag, void* userdata) {
+    auto password = *reinterpret_cast<const std::string*>(userdata);
+    std::size_t copy_size =
+        std::min(static_cast<std::size_t>(std::max(buffer_size, 0)), password.size());
+    if (copy_size < password.size()) {
+        return -1;
+    }
+    memcpy(buffer, password.c_str(), copy_size);
+    return static_cast<int>(copy_size);
+}
+
+auto open_file(const std::string& client_cert_fn) {
+    FILE* p = fopen(client_cert_fn.c_str(), "r");
+    if (p == nullptr) {
+        std::ostringstream what;
+        what << "Can't open file '" << client_cert_fn << "'";
+        raise(what.str(), strerror(errno));
+    }
+    auto close_file = [](FILE* p) { if (p) { fclose(p); } };
+    return std::unique_ptr<FILE, decltype(close_file)>(p, close_file);
+}
+
+void raise(const std::string& client_cert_fn, const std::string& entry, unsigned long ssl_error) {
+    std::ostringstream what;
+    what << "Can't read the " << entry << " from the '" << client_cert_fn << "' file";
+
+    std::ostringstream reason;
+    // The functions `PEM_read_PrivateKey` and `PEM_read_X509` set the `PEM_R_BAD_PASSWORD_READ`
+    // error code if the callback returned `-1`.  This can only happen if the password is too long
+    // (please see the `passwd_cb` function). Check that specific condition in order to avoid
+    // too vague "bad password read" error message.
+    if (ERR_GET_LIB(ssl_error) == ERR_LIB_PEM &&
+        ERR_GET_REASON(ssl_error) == PEM_R_BAD_PASSWORD_READ) {
+        reason << "password is too long. Please reencrypt the " <<  entry
+               << " with a shorter password.";
+    } else {
+        reason << ERR_reason_error_string(ssl_error);
+    }
+
+    raise(what.str(), reason.str());
+}
+
+auto read_private_key(const std::string& client_cert_fn,
+                      const std::string& client_cert_passwd) {
+    auto cb_data = const_cast<std::string*>(&client_cert_passwd);
+    EVP_PKEY* key =
+        PEM_read_PrivateKey(open_file(client_cert_fn).get(), nullptr, passwd_cb, cb_data);
+    if (key == nullptr) {
+        raise(client_cert_fn, "private key", ERR_peek_last_error());
+    }
+    auto key_deleter = [](EVP_PKEY* p) { if (p) { EVP_PKEY_free(p); } };
+    return std::unique_ptr<EVP_PKEY, decltype(key_deleter)> (key, key_deleter);
+}
+
+struct cert_deleter {
```

Comment:
This is mostly matter of style/taste but I would consider using plain function instead of struct for this deleter object.